### PR TITLE
Add net, tcp, uptime and cpu metrics

### PR DESCRIPTION
**Pull request by logarithm**

### CODE REVIEW

File: src/main/java/com/metamx/metrics/SysMonitor.java [hunk 1]
```diff
@@ -75,7 +77,8 @@ public SysMonitor(Map<String, String[]> dimensions)
             new DiskStats(),
             new NetStats(),
             new CpuStats(),
-            new SwapStats()
```
Comment (xvrl) [inline src/main/java/com/metamx/metrics/SysMonitor.java]: why are we removing swap stats?

File: src/main/java/com/metamx/metrics/SysMonitor.java [hunk 2]
```diff
@@ -141,7 +144,9 @@ public void emit(ServiceEmitter emitter)
       if (mem != null) {
         final Map<String, Long> stats = ImmutableMap.of(
             "sys/mem/max", mem.getTotal(),
-            "sys/mem/used", mem.getUsed()
+            "sys/mem/used", mem.getUsed(),
+            "sys/mem/actual/used", mem.getActualUsed(),
```
Comment (xvrl) [inline src/main/java/com/metamx/metrics/SysMonitor.java]: can you explain the difference with the existing ones?

File: src/main/java/com/metamx/metrics/SysMonitor.java [hunk 3]
```diff
@@ -268,10 +273,16 @@ public void emit(ServiceEmitter emitter)
...
-                  "sys/fs/max", fsu.getTotal() * 1024,
-                  "sys/fs/used", fsu.getUsed() * 1024
-              );
+              final Map<String, Long> stats = ImmutableMap.<String, Long>builder()
+                  .put("sys/fs/max", fsu.getTotal() * 1024)
+                  .put("sys/fs/used", fsu.getTotal() * 1024)
+                  .put("sys/fs/files/count", fsu.getFiles())
+                  .put("sys/fs/files/free", fsu.getFreeFiles())
+                  .put("sys/fs/disk/read/count", fsu.getDiskReads())
+                  .put("sys/fs/disk/read/size", fsu.getDiskReadBytes())
```
Comment (xvrl) [inline src/main/java/com/metamx/metrics/SysMonitor.java]: how are those metrics different than the existing `sys/disk/(read|write)/(size|count)`

File: src/main/java/com/metamx/metrics/SysMonitor.java [hunk 4]
```diff
@@ -382,10 +393,21 @@ public void emit(ServiceEmitter emitter)
...
+                      .put("sys/net/read/dropped", netstat.getRxDropped())
+                      .put("sys/net/read/overruns", netstat.getRxOverruns())
+                      .put("sys/net/read/frame", netstat.getRxFrame())
+                      .put("sys/net/write/size", netstat.getTxBytes())
+                      .put("sys/net/write/packets", netstat.getTxPackets())
+                      .put("sys/net/write/errors", netstat.getTxErrors())
+                      .put("sys/net/write/dropped", netstat.getTxDropped())
+                      .put("sys/net/write/collisions", netstat.getTxCollisions())
+                      .put("sys/net/write/overruns", netstat.getTxOverruns())
+                      .put("sys/net/write/carrier", netstat.getTxCarrier())
```
Comment (xvrl) [inline src/main/java/com/metamx/metrics/SysMonitor.java]: what's carrier?

File: src/main/java/com/metamx/metrics/SysMonitor.java (hunk 1 above)
Comment (logarithm) [inline src/main/java/com/metamx/metrics/SysMonitor.java]: Sorry, my bad

File: src/main/java/com/metamx/metrics/SysMonitor.java (hunk 2 above)
Comment (logarithm) [inline src/main/java/com/metamx/metrics/SysMonitor.java]: From sigar source code:
```
getUsed - Total used system memory
getActualUsed - Actual total used system memory (e.g. Linux minus buffers)
```

File: src/main/java/com/metamx/metrics/SysMonitor.java (hunk 3 above)
Comment (logarithm) [inline src/main/java/com/metamx/metrics/SysMonitor.java]: I have checked it now - it's the same metrics. Will remove it

File: src/main/java/com/metamx/metrics/SysMonitor.java [hunk 5]
```diff
@@ -382,10 +394,21 @@ public void emit(ServiceEmitter emitter)
...
```
Comment (logarithm) [inline src/main/java/com/metamx/metrics/SysMonitor.java]: @xvrl, have lost your comment

(http://www.linuxquestions.org/questions/linux-networking-3/what-is-carrier-from-the-ifconfig-output-57333/#post293740)

Not sure, that it is a useful metric

File: src/main/java/com/metamx/metrics/SysMonitor.java [hunk 6]
```diff
@@ -268,10 +275,12 @@ public void emit(ServiceEmitter emitter)
               log.error(e, "Failed to get FileSystemUsage[%s]", name);
             }
             if (fsu != null) {
-              final Map<String, Long> stats = ImmutableMap.of(
-                  "sys/fs/max", fsu.getTotal() * 1024,
-                  "sys/fs/used", fsu.getUsed() * 1024
-              );
+              final Map<String, Long> stats = ImmutableMap.<String, Long>builder()
```
Comment (xdralex) [inline src/main/java/com/metamx/metrics/SysMonitor.java]: I think stats like diskQueue/diskServiceTime are also worth to add (you can find them in FileSystemUsage or DiskUsage, probably adding them under disk umbrella makes more sense).

File: src/main/java/com/metamx/metrics/SysMonitor.java [hunk 7]
```diff
@@ -449,4 +471,100 @@ public void emit(ServiceEmitter emitter)
...
+      Uptime uptime = null;
+      try {
+        uptime = sigar.getUptime();
+      }
+      catch (SigarException e) {
+        log.error(e, "Failed to get Uptime");
+      }
+      if (uptime != null) {
+        final Map<String, Long> stats = ImmutableMap.of(
+            "sys/uptime", Double.valueOf(uptime.getUptime()).longValue()
```
Comment (xdralex) [inline src/main/java/com/metamx/metrics/SysMonitor.java]: Do you mind adding getLoadAverage? I think all three l/a of 1/5/15 min make sense.

**Comment by xdralex** [comment]
Looks awesome to me! Let's add couple more gauges and feel free to merge it.

File: src/main/java/com/metamx/metrics/SysMonitor.java (hunk 2 above)
Comment (drcrallen) [inline src/main/java/com/metamx/metrics/SysMonitor.java]: Buffers are almost never used compared to page cache. I don't think those will be different.

**Pull request merged**